src/config.js

```diff
@@ -2,23 +2,89 @@
  * #### Import members from **@edx/frontend-platform**
  *
  * The configuration module provides utilities for working with an application's configuration
- * document (ConfigDocument).  This module uses `process.env` to import configuration variables
- * from the command-line build process.  It can be dynamically extended at run-time using a
- * `config` initialization handler.  Please see the Initialization documentation for more
- * information on handlers and initialization phases.
+ * document (ConfigDocument).  Configuration environment variables can be supplied to the
+ * application in four different ways:
+ *
+ * ##### Build-time Environment Variables
+ *
+ * A set list of required config variables can be supplied as
+ * command-line environment variables during the build process.
+ *
+ * As a simple example, these are supplied on the command-line before invoking `npm run build`:
  *
  * ```
- * import { getConfig } from '@edx/frontend-platform';
+ * LMS_BASE_URL=http://localhost:18000 npm run build
+ * ```
  *
- * const {
- *   BASE_URL,
- *   LMS_BASE_URL,
- *   LOGIN_URL,
- *   LOGIN_URL,
- *   REFRESH_ACCESS_TOKEN_ENDPOINT,
- *   ACCESS_TOKEN_COOKIE_NAME,
- *   CSRF_TOKEN_API_PATH,
- * } = getConfig();
+ * Note that additional variables _cannot_ be supplied via this method without using the `config`
+ * initialization handler.  The app won't pick them up and they'll appear `undefined`.
+ *
+ * ##### JavaScript File Configuration
+ *
+ * Configuration variables can be supplied in an optional file
+ * named env.config.js.  This file must export either an Object containing configuration variables
+ * or a function.  The function must return an Object containing configuration variables or,
+ * alternately, a promise which resolves to an Object.
+ *
+ * Exporting a config object:
+ * ```
+ * const config = {
+ *   LMS_BASE_URL: 'http://localhost:18000'
+ * };
+ *
+ * export default config;
+ * ```
+ *
+ * Exporting a function that returns an object:
+ * ```
+ * function getConfig() {
+ *   return {
+ *     LMS_BASE_URL: 'http://localhost:18000'
+ *   };
+ * }
+ * ```
+ *
+ * Exporting a function that returns a promise that resolves to an object:
+ * ```
+ * function getAsyncConfig() {
+ *   return new Promise((resolve, reject) => {
+ *     resolve({
+ *       LMS_BASE_URL: 'http://localhost:18000'
+ *     });
+ *   });
+ * }
+ *
+ * export default getAsyncConfig;
+ * ```
+ *
+ * ##### Runtime Configuration
```

Comment:
could be nice to add something explaining how this interacts with https://github.com/openedx/frontend-platform/pull/474/files#diff-23821ae067a9c7dfb6227774aefb39fa3951a5874ea953b9401dfc80175c026cR257-R259
> When using Webpack (i.e., normal usage), the build process is responsible for supplying these
> variables via command-line environment variables.  That means they must be supplied at build
> time.

Reply:
Yes!  These are good thoughts.  There's an order that they get applied that I don't think we've called out anywhere.

Reply:
I added some words around this and clarified why you'd use one or the other.  Resolving!